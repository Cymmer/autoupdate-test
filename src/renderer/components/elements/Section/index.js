import React from 'react';
import cn from 'classnames';
import PropTypes from 'prop-types';
import GLOBALS from 'codechum-app-globals';
import styles from './styles.module.scss';

import Text from '../Text';
import textTypes from '../Text/constants/textTypes';

const Section = ({ id, children, className, title }) => (
  <section className={cn(styles.Section, className)} id={id}>
    {title && (
      <Text
        className={styles.Section_title}
        colorClass={GLOBALS.COLOR_CLASSES.NEUTRAL['400']}
        type={textTypes.HEADING.XXXS}
      >
        {title}
      </Text>
    )}
    {children}
  </section>
);

Section.defaultProps = {
  id: null,
  className: null,
  children: null,
  title: null,
};

Section.propTypes = {
  id: PropTypes.string,
  className: PropTypes.string,
  children: PropTypes.any,
  title: PropTypes.string,
};

export default Section;
