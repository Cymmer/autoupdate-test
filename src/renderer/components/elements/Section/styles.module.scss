@import '../../../scss/variables';

.Section {
  margin: 48px 0 64px;

  &_title {
    margin: 0 0 24px 0;
  }
}
