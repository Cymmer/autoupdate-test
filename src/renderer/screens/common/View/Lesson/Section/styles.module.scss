@import '../../../../../scss/variables';
@import '../../../../../scss/themes';

.Section {
  margin: 80px 0;

  &_title {
    margin: 0 0 16px 0;
  }

  > p {
    margin: 0 0 16px 0;

    > code {
      color: $red-300;
    }

    > a {
      color: $blue-300;
    }
  }

  > div > pre {
    margin: 16px 0 !important;
  }

  &:last-child {
    margin: 80px 0 120px;
  }
}
