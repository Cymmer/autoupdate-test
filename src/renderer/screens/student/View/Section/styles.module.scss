@import '../../../../scss/variables';
@import '../../../../scss/mixins';

@mixin status($color) {
  margin: 0 0 0 16px;
  @include flex-center;

  &:before {
    content: '';
    width: 12px;
    height: 12px;
    background-color: $color;
    border-radius: 100%;
    display: block;
    margin: 0 8px 0 0;
  }
}

.ViewSection {
  &_header {
    @include flex-between;
    align-items: center;
    width: 100%;
  }

  &_image {
    width: 80px;
    height: 80px;
    background-size: 48px;
    border-radius: 16px;

    @include breakpoint(sm) {
      margin: 0 0 16px 0;
      width: 100%;
      height: 96px;
      background-size: 96px;
    }
  }

  &_info {
    @include flex-center;

    @include breakpoint(sm) {
      flex-direction: column;
      width: 100%;
      align-items: flex-start;
    }
  }

  &_name {
    display: flex;
    align-items: center;
    flex-wrap: wrap;
  }

  &_text {
    margin: 0 0 0 24px;

    @include breakpoint(sm) {
      flex-direction: column;
      align-items: flex-start;
      margin: 0;
    }
  }

  &_status {
    &___active {
      @include status($green-300);
    }

    &___inactive {
      @include status($red-300);
    }
  }

  &_download {
    > i {
      color: $yellow-300;
    }

    &:disabled {
      @include theme {
        color: t('textColor-300');
        background-color: t('textColor-50');
      }

      > i {
        @include theme {
          color: t('textColor-300');
        }
      }
    }

    @include breakpoint(sm) {
      display: none;
    }
  }
}
