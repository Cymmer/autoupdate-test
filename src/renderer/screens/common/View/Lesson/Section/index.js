import React from 'react';
import PropTypes from 'prop-types';
import GLOBALS from 'codechum-app-globals';
import styles from './styles.module.scss';

import { Text } from '../../../../../components/elements';
import { textTypes } from '../../../../../components/elements/constants';

const Section = ({ title, children }) => (
  <div className={styles.Section}>
    {title && (
      <Text
        className={styles.Section_title}
        colorClass={GLOBALS.COLOR_CLASSES.NEUTRAL['400']}
        type={textTypes.HEADING.XS}
      >
        {title}
      </Text>
    )}
    {children}
  </div>
);

Section.defaultProps = {
  title: null,
  children: null,
};

Section.propTypes = {
  title: PropTypes.string,
  children: PropTypes.any,
};

export default Section;
