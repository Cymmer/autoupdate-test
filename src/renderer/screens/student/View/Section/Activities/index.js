import GLOBALS from 'codechum-app-globals';
import { TaskCardLargeV4Student, TaskList } from 'components';
import { NoResults, Section } from 'components/elements';
import { noResultsTypes } from 'components/elements/constants';
import { getUser } from 'ducks';
import { useStudentTasks, useTimedTasks } from 'hooks';
import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';

export const Activities = ({
  section: { id: sectionId },
  user: { id: studentId },
}) => {
  // ongoing and future tasks
  const { isLoading: ongoingTasksIsLoading, data: ongoingTasks } =
    useStudentTasks({
      studentId,
      params: {
        status: 'present,future',
        sectionId,
        isCourseActivity: false,
      },
    });

  const { runningTasks: runningOngoingTasks } = useTimedTasks(ongoingTasks);

  return (
    <Section id="ongoingActivitiesStudentSection" title="Ongoing Activities">
      <TaskList isLoading={ongoingTasksIsLoading}>
        {!ongoingTasksIsLoading && runningOngoingTasks?.length === 0 ? (
          <NoResults
            codyType={GLOBALS.CODY_TYPES.OK}
            text="Looks like you don't have any activities to do"
            type={noResultsTypes.SMALL}
          />
        ) : (
          runningOngoingTasks?.map((taskWithDetails) => (
            <TaskCardLargeV4Student
              key={taskWithDetails.id}
              task={taskWithDetails}
            />
          ))
        )}
      </TaskList>
    </Section>
  );
};

Activities.propTypes = {
  section: PropTypes.object.isRequired,
  user: PropTypes.object.isRequired,
};

const mapStateToProps = (store) => ({
  user: getUser(store),
});

export default connect(mapStateToProps, null)(Activities);
