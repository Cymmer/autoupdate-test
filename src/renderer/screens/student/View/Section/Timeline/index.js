import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';
import { connect } from 'react-redux';
import { useHistory } from 'react-router-dom';

import { SectionCourseTimelineV4 } from 'components';

import {
  useResults,
  useLessonObjectStatusExist,
  useCourseDataV4,
  useMixpanelTrack,
} from 'hooks';
import {
  isCourseQuiz,
  isTaskFinished,
  isTaskHasntStarted,
  momentToDateTimeWithTimezone,
  isSameOrAfterWithoutSeconds,
} from 'codechum-app-utils';
import { actionTypes } from 'components/elements/constants';
import { getUser } from 'ducks';
import { LessonObjectStatusesService } from 'services';
import GLOBALS from 'codechum-app-globals';
import { viewSectionTabs } from '../../../constants';
import Preloader from '../../Preloader';
import styles from './styles.module.scss';

const { CourseObjectCard, NextCourseObjectCard } = SectionCourseTimelineV4;

const determineObjectValues = ({ courseObjects }) => {
  // we remove the introduction and the summary objects
  const filteredObjects = courseObjects.filter(
    (obj) => !obj?.is_introduction && !obj?.is_summary
  );

  const unlockedCourseObjects = filteredObjects.filter((obj, index) =>
    isCourseQuiz(obj)
      ? obj.is_locked === false &&
        filteredObjects[index - 1]?.has_finished_previous_lesson === true
      : obj.is_locked === false && obj.has_finished_previous_lesson === true
  );
  const lockedCourseObjects = filteredObjects.filter((obj, index) =>
    isCourseQuiz(obj)
      ? obj.is_locked === true ||
        filteredObjects[index - 1]?.has_finished_previous_lesson === false
      : obj.is_locked === true || obj.has_finished_previous_lesson === false
  );

  // we remove the last unlocked course object from the unlocked courses
  // and set it as the next course object
  const nextCourseObject = unlockedCourseObjects.pop();

  return {
    unlockedCourseObjects,
    nextCourseObject,
    lockedCourseObjects,
  };
};

const determineQuizCardText = (task, results) => {
  if (isTaskFinished(task)) {
    return 'Review Quiz';
  }

  // if it went here, it means that the task is already starting

  const taskResult = results.find((result) => result.task_id === task.id);

  let isRetakeFinished = false;

  if (taskResult?.retake_end_datetime) {
    const currTime = moment();
    const endTime = moment(taskResult.retake_end_datetime);
    isRetakeFinished = isSameOrAfterWithoutSeconds(currTime, endTime);
  }

  if (taskResult?.datetime_submitted || isRetakeFinished) {
    return 'Review Quiz';
  }

  if (taskResult) {
    return 'Continue Quiz';
  }

  return 'Start Quiz';
};

const determineQuizCardLink = (task, results) => {
  if (isTaskFinished(task)) {
    return `/student/review/${task.id}/question/1`;
  }

  // if it went here, it means that the task is already starting

  const taskResult = results.find((result) => result.task_id === task.id);

  let isRetakeFinished = false;

  if (taskResult?.retake_end_datetime) {
    const currTime = moment();
    const endTime = moment(taskResult.retake_end_datetime);
    isRetakeFinished = isSameOrAfterWithoutSeconds(currTime, endTime);
  }

  if (taskResult?.datetime_submitted || isRetakeFinished) {
    return `/student/review/${task.id}/question/1`;
  }

  return `/student/answer/${task.id}/question/1`;
};

const determineQuizCardIsDisabled = (task) => {
  if (isTaskFinished(task)) return task.has_exam_ongoing;
  if (task.has_submitted) return true;

  return isTaskHasntStarted(task);
};

const Timeline = ({
  courseDataId,
  section: { id: sectionId },
  user: { id: studentId },
  downloadCertificateButton,
}) => {
  const history = useHistory();

  const { isLoading: isCourseDataLoading, courseData } = useCourseDataV4({
    courseDataId,
  });

  const { course_objects: courseObjects } = courseData || {};
  const { isLoading: areResultsLoading, results } = useResults({
    isExtended: false,
    params: {
      taskIds: courseObjects
        ?.filter((obj) => isCourseQuiz(obj))
        ?.map((courseObject) => courseObject.task.id)
        .join(','),
      studentId,
    },
    isPaused: courseObjects == null,
  });

  const {
    isLoading: doesLessonObjectStatusExistLoading,
    doesExist: doesIntroductionLessonObjectStatusExist,
  } = useLessonObjectStatusExist({
    params: {
      lessonObjectId: courseObjects
        ? courseObjects.find((obj) => obj.is_introduction).first_lesson_object
            .id
        : null,
      studentId,
    },
  });

  useMixpanelTrack({
    body: {
      event_name: 'Go to View Class Screen > Timeline Tab',
      event_properties: {
        'Section ID': sectionId,
      },
    },
  });

  if (
    isCourseDataLoading ||
    areResultsLoading ||
    doesLessonObjectStatusExistLoading
  ) {
    return <Preloader />;
  }

  const introductionObject = courseObjects.find(
    (obj) => obj.is_introduction === true
  );

  const { unlockedCourseObjects, nextCourseObject, lockedCourseObjects } =
    determineObjectValues({ courseObjects });

  return (
    <>
      <SectionCourseTimelineV4
        data-test="sectionCourseTimeline"
        introductionLink={`/student/classes/${sectionId}/${viewSectionTabs.TIMELINE.value}/lessons/${introductionObject.slug}/${introductionObject.first_lesson_object.slug}`}
        lockedCourseObjectsJsx={lockedCourseObjects.map(
          (courseObject, index) => (
            <CourseObjectCard
              key={courseObject.id}
              actions={
                isCourseQuiz(courseObject)
                  ? [
                      {
                        type: actionTypes.BUTTON,
                        text: 'Start Quiz',
                        disabled: courseObject.is_locked,
                        action:
                          courseObject.is_locked === false
                            ? () => {
                                history.push(
                                  determineQuizCardLink(
                                    courseObject.task,
                                    results
                                  )
                                );
                              }
                            : () => {},
                      },
                    ]
                  : [
                      {
                        type: actionTypes.BUTTON,
                        text: 'View Lesson',
                        disabled: true,
                        tooltip:
                          courseObject.is_locked === false &&
                          courseObject.has_finished_previous_lesson === false
                            ? 'Your teacher has already unlocked this but you need to finish the previous lesson in order to view this one.'
                            : null,
                        action: () => {},
                      },
                    ]
              }
              courseObject={courseObject}
              hasEvenPosition={index % 2 === 0}
            />
          )
        )}
        nextCourseObjectJsx={
          nextCourseObject && (
            <NextCourseObjectCard
              actions={
                isCourseQuiz(nextCourseObject)
                  ? [
                      {
                        type: actionTypes.BUTTON,
                        text: determineQuizCardText(
                          nextCourseObject.task,
                          results
                        ),
                        disabled: determineQuizCardIsDisabled(
                          nextCourseObject.task
                        ),
                        action: () => {
                          history.push(
                            determineQuizCardLink(
                              nextCourseObject.task,
                              results
                            )
                          );
                        },
                      },
                    ]
                  : [
                      {
                        type: actionTypes.BUTTON,
                        text: 'View Lesson',
                        disabled: false,
                        action: () => {
                          if (!doesIntroductionLessonObjectStatusExist) {
                            // before redirecting to the Introduction topic, create a
                            // LessonObjectStatus first
                            LessonObjectStatusesService.create({
                              body: {
                                datetime_started: momentToDateTimeWithTimezone(
                                  moment()
                                ),
                                lesson_object_id:
                                  introductionObject.first_lesson_object.id,
                                student_id: studentId,
                              },
                            }).then(() => {
                              history.push(
                                `/student/classes/${sectionId}/${viewSectionTabs.TIMELINE.value}/lessons/${introductionObject.slug}/${introductionObject.first_lesson_object.slug}`
                              );
                            });
                          } else {
                            history.push(
                              `/student/classes/${sectionId}/${viewSectionTabs.TIMELINE.value}/lessons/${nextCourseObject.slug}/${nextCourseObject.first_lesson_object.slug}`
                            );
                          }
                        },
                      },
                    ]
              }
              courseObject={nextCourseObject}
              userType={GLOBALS.USER_TYPES.STUDENT}
            />
          )
        }
        summaryLink="#"
        unlockedCourseObjectsJsx={unlockedCourseObjects.map(
          (courseObject, index) => (
            <CourseObjectCard
              key={courseObject.id}
              actions={
                isCourseQuiz(courseObject)
                  ? [
                      {
                        type: actionTypes.BUTTON,
                        text: determineQuizCardText(courseObject.task, results),
                        disabled: determineQuizCardIsDisabled(
                          courseObject.task
                        ),
                        action: () => {
                          history.push(
                            determineQuizCardLink(courseObject.task, results)
                          );
                        },
                      },
                    ]
                  : [
                      {
                        type: actionTypes.BUTTON,
                        text: 'View Lesson',
                        disabled: false,
                        action: () => {
                          history.push(
                            `/student/classes/${sectionId}/${viewSectionTabs.TIMELINE.value}/lessons/${courseObject.slug}/${courseObject.first_lesson_object.slug}`
                          );
                        },
                      },
                    ]
              }
              courseObject={courseObject}
              hasEvenPosition={index % 2 === 0}
              userType={GLOBALS.USER_TYPES.STUDENT}
            />
          )
        )}
      />
      {downloadCertificateButton && (
        <div
          className={styles.Timeline_buttonContainer}
          data-test="downloadCertificateButtonContainer"
        >
          {downloadCertificateButton}
        </div>
      )}
    </>
  );
};

Timeline.defaultProps = {
  downloadCertificateButton: null,
};

Timeline.propTypes = {
  courseDataId: PropTypes.number.isRequired,
  user: PropTypes.object.isRequired,
  section: PropTypes.object.isRequired,
  downloadCertificateButton: PropTypes.node,
};

const mapStateToProps = (store) => ({
  user: getUser(store),
});

export default connect(mapStateToProps, null)(Timeline);
