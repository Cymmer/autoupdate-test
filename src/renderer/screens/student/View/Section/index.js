import GLOBALS from 'codechum-app-globals';
import { formatSectionSchedule } from 'codechum-app-utils';
import { SectionImage } from 'components';
import {
  sectionColors,
  sectionIcons,
  sectionListTypes
} from 'components/constants';
import { Container, Header, SubInfo, Text } from 'components/elements';
import { tabKinds, textTypes } from 'components/elements/constants';
import { useSectionV4 } from 'hooks';
import PropTypes from 'prop-types';
import React from 'react';
import { Redirect, Route, Switch } from 'react-router-dom';
import { STUDENT_ROUTES } from 'screen-wrappers/Student/constants';
import viewSectionTabs from '../constants/viewSectionTabs';
import Preloader from '../Preloader';
import Activities from './Activities';
import styles from './styles.module.scss';
import Timeline from './Timeline';

export const ViewSection = ({
  match: {
    params: { sectionId, activeTab },
  },
}) => {
  const { isLoading: isSectionLoading, section } = useSectionV4({
    isExtended: true,
    sectionId,
  });

  const {
    name: sectionName,
    schedules,
    is_active: isActive,
    icon,
    color,
    // teacher_id: teacherId,
    course_data_id: courseDataId,
  } = section || {};

  // const { isLoading: isTeacherLoading, user: teacher } = useUser({
  //   isExtended: false,
  //   userId: teacherId,
  //   params: {
  //     teachers: true,
  //     managers: true,
  //   },
  // });

  if (isSectionLoading) {
    return <Preloader />;
  }

  return (
    <>
      <Header
        activeTab={activeTab}
        breadcrumbs={[
          {
            name: 'My Classes',
            link: '/student/classes',
          },
          {
            name: sectionName,
            link: '#',
          },
        ]}
        tabs={[
          courseDataId
            ? {
                name: viewSectionTabs.TIMELINE.name,
                value: viewSectionTabs.TIMELINE.value,
                kind: tabKinds.LINK,
                action: `/student/classes/${sectionId}/${viewSectionTabs.TIMELINE.value}`,
              }
            : null,
          {
            name: viewSectionTabs.ACTIVITIES.name,
            value: viewSectionTabs.ACTIVITIES.value,
            kind: tabKinds.LINK,
            action: `/student/classes/${sectionId}/${viewSectionTabs.ACTIVITIES.value}`,
          },
        ].filter((tab) => tab !== null)}
      >
        <div className={styles.ViewSection_header}>
          <div className={styles.ViewSection_info}>
            <SectionImage
              className={styles.ViewSection_image}
              color={
                color
                  ? sectionColors[
                      [...Object.keys(sectionColors)].find(
                        (c) => sectionColors[c].value === color
                      )
                    ]
                  : sectionColors.GREEN
              }
              icon={
                icon
                  ? sectionIcons[
                      [...Object.keys(sectionIcons)].find(
                        (i) => sectionIcons[i].value === icon
                      )
                    ]
                  : sectionIcons.CODY
              }
              size={sectionListTypes.SMALL}
            />

            <div className={styles.ViewSection_text}>
              <div className={styles.ViewSection_name}>
                <Text type={textTypes.HEADING.SM}>{sectionName}</Text>
                <Text
                  className={
                    styles[
                      `ViewSection_status___${isActive ? 'active' : 'inactive'}`
                    ]
                  }
                  colorClass={GLOBALS.COLOR_CLASSES.NEUTRAL['400']}
                  type={textTypes.BODY.SM}
                >
                  {isActive ? 'Active' : 'Inactive'}
                </Text>
              </div>
              <div className={styles.ViewSection_subinfo}>
                <SubInfo
                  data={[
                    // {
                    //   icon: 'person',
                    //   text: formatTeacherName(teacher),
                    // },
                    {
                      icon: 'access_time',
                      text: formatSectionSchedule(schedules),
                    },
                  ]}
                />
              </div>
            </div>
          </div>
        </div>
      </Header>
      <Container>
        <React.Suspense fallback={<div>loading</div>}>
          <Switch>
            {/*
              We do a special conditional redirect to the section's class overview for
              the student side because the `enroll` endpoint doesn't return the section
              enrolled so we always redirect the student from the Join Section screen
              to the section's timeline because we don't know if the section that the
              student just joined has a course data or not
            */}
            {courseDataId ? (
              <Route
                name="Class - Timeline"
                path={`/student/classes/:sectionId/${viewSectionTabs.TIMELINE.value}`}
                render={(props) => (
                  <Timeline
                    courseDataId={courseDataId}
                    section={section}
                    {...props}
                  />
                )}
              />
            ) : (
              <Route
                name="Class - Activities"
                path={`/student/classes/:sectionId/${viewSectionTabs.TIMELINE.value}`}
                render={() => (
                  <Redirect
                    to={`/student/classes/${sectionId}/${viewSectionTabs.ACTIVITIES.value}`}
                  />
                )}
              />
            )}
            <Route
              name="Class - Activities"
              path={`/student/classes/:sectionId/${viewSectionTabs.ACTIVITIES.value}`}
              render={(props) => <Activities section={section} {...props} />}
            />

            <Redirect to={STUDENT_ROUTES.VIEW_CLASS} />
          </Switch>
        </React.Suspense>
      </Container>
    </>
  );
};

ViewSection.propTypes = {
  match: PropTypes.object.isRequired,
};

export default ViewSection;
