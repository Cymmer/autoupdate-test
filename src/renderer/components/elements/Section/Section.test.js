import React from 'react';
import { shallow } from 'enzyme';
import Section from './index';

describe('RENDERING - Section', () => {
  it('should render with default props', () => {
    const component = shallow(<Section />);
    expect(component).toMatchSnapshot();
  });

  it('should render its children if it has children', () => {
    const component = shallow(<Section>Test</Section>);
    expect(component).toMatchSnapshot();
  });

  it('should render a title if it has a title', () => {
    const component = shallow(<Section title="Test" />);
    expect(component).toMatchSnapshot();
  });
});
