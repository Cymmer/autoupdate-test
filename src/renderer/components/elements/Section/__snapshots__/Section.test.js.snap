// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RENDERING - Section should render a title if it has a title 1`] = `
<section
  className="Section"
  id={null}
>
  <Text
    className="Section_title"
    colorClass="n___400"
    element={null}
    id={null}
    type="heading___xxxs"
  >
    Test
  </Text>
</section>
`;

exports[`RENDERING - Section should render its children if it has children 1`] = `
<section
  className="Section"
  id={null}
>
  Test
</section>
`;

exports[`RENDERING - Section should render with default props 1`] = `
<section
  className="Section"
  id={null}
/>
`;
